rest-api/offline/table_exporter.py

```diff
@@ -14,20 +26,89 @@ class TableExporter(object):
   other CSV clients (e.g. BigQuery, Google Sheets) can actually understand.
   """
 
+  @staticmethod
+  def _obfuscate_id(pmi_id, salt):
+    """
+    One-way ID obfuscation, intended for PMI participant IDs.
+
+    This method should aim to (1) avoid hash collisions across PMI ID inputs and (2) not be
+    reversible or rerunnable by anyone without RDR production data access.
+
+    Params:
+      - pmi_id (number) must be an integer participant ID
+      - salt (number) must be a long
+
+    Returns: a positive integer
+    """
+    h = hashlib.sha1()
+    h.update(struct.pack('>l', pmi_id))
+    h.update(struct.pack('>q', salt))
+    # Just take the first 4 bytes so that the output ID is an integer, roughly in the same domain as
+    # the input PMI participant ID.
+    b = h.digest()[0:4]
+    return abs(struct.unpack('>l', b)[0])
+
   @classmethod
-  def _export_csv(cls, bucket_name, database, directory, table_name):
+  def _export_csv(cls, bucket_name, database, directory, deidentify_salt, table_name):
     assert _TABLE_PATTERN.match(table_name)
     assert _TABLE_PATTERN.match(database)
-    SqlExporter(bucket_name, use_unicode=True).run_export('%s/%s.csv' % (directory, table_name),
-                                                          'SELECT * FROM %s.%s' %
-                                                          (database, table_name))
+
+    transformf = None
+    if deidentify_salt:
+      # Deidentification requested: hash outgoing participant IDs with a consistent salt across this
+      # export. Cache obfuscated participant IDs across row callbacks to avoid recomputation and to
+      # detect collisions.
+      pmi_to_obfuscated = {}
+      obfuscated_to_pmi = {}
+      def f(row_proxy):
+        out = [v for v in row_proxy]
+        for i, key in enumerate(row_proxy.keys()):
+          if key != 'participant_id':
+            continue
+
+          pmi_id = out[i]
+          if pmi_id not in pmi_to_obfuscated:
+            obf_id = TableExporter._obfuscate_id(pmi_id, deidentify_salt)
+            pmi_to_obfuscated[pmi_id] = obf_id
```

Comment:
Passing this function around having it share the to/from mapping state ... closures are so lovely.